packages/ifpack2/src/Ifpack2_BlockTriDiContainer_def.hpp

```diff
@@ -0,0 +1,2756 @@
+    LOList A_colindsub;
+    // Tridiag block values. pack_td_ptr(i) points to the start of the i'th
+    // tridiag's pack, and i % vector_length gives the position in the pack.
+    Values values;
+
+    // Index into row-major block of a tridiag.
+    template <typename Int> static KOKKOS_INLINE_FUNCTION
+    Int ind2row (const Int& ind) { return (ind + 1) / 3; }
+    // Given a row of a row-major tridiag, return the index of the first block
+    // in that row.
+    template <typename Int> static KOKKOS_INLINE_FUNCTION
+    Int row2ind (const Int& row) { return row > 0 ? 3*row - 1 : 0; }
+    // Number of blocks in a tridiag having a given number of rows.
+    template <typename Int> static KOKKOS_INLINE_FUNCTION
+    Int nblks (const Int& nrows) { return nrows > 0 ? 3*nrows - 2 : 0; }
+  };
+
+  // A - Tridiags(A), i.e., R in the splitting A = D + R.
+  struct AmD {
+    // rowptr points to the start of each row of A_colindsub.
+    SizeList rowptr, rowptr_remote;
+    // Indices into A's rows giving the blocks to extract. rowptr(i) points to
+    // the i'th row. Thus, g.entries(A_colindsub(rowptr(row) : rowptr(row+1))),
+    // where g is A's graph, are the columns AmD uses. If seq_method_, then
+    // A_colindsub contains all the LIDs and A_colindsub_remote is empty. If !
+    // seq_method_, then A_colindsub contains owned LIDs and A_colindsub_remote
+    // contains the remote ones.
+    LOList A_colindsub, A_colindsub_remote;
+
+    // Currently always true.
+    bool is_tpetra_block_crs;
+    // If is_tpetra_block_crs, then this is a pointer to A_'s value data.
+    BlockCrsValues tpetra_values;
+  };
+
+public: // for Cuda
+
+  class AsyncableImport;
+  class NormManager;
+
+private:
+  BlockTriDiContainer<MatrixType, LocalScalarType>& container_;
+  Teuchos::RCP<const block_crs_matrix_type> A_bcrs_;
+  // Use the slow, straightforward import(dm->cm), matvec, solve pattern?
+  bool seq_method_;
+
+  // Some terms:
+  //   The matrix A is split as A = D + R, where D is the matrix of tridiag
+  // blocks and R is the remainder.
+  //   A part is roughly a synonym for a tridiag. The distinction is that a part
+  // is the set of rows belonging to one tridiag and, equivalently, the off-diag
+  // rows in R associated with that tridiag. In contrast, the term tridiag is
+  // used to refer specifically to tridiag data, such as the pointer into the
+  // tridiag data array.
+  //   Local (lcl) row arge the LIDs. lclrow lists the LIDs belonging to each
+  // tridiag, and partptr points to the beginning of each tridiag. This is the
+  // LID space.
+  //   Row index (idx) is the ordinal in the tridiag ordering. lclrow is indexed
+  // by this ordinal. This is the 'index' space.
+  //   A flat index is the mathematical index into an array. A pack index
+  // accounts for SIMD packing.
+
+  // Local row LIDs. Permutation from caller's index space to tridiag index
+  // space.
+  LOList lclrow_;
```

Comment:
Bunch of LOL :)

Reply:
@srajama1, SplitJacobi is fine. What do you want the corresponding ParameterList entry to be for "relaxation: type"? Currently it's "MT Jacobi".

Reply:
MT Split Jacobi ?